[Ng Xing Yu] iP

ngxingyu opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
---
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

ngxingyu committed:
Implement Greet Echo Exit

ngxingyu committed:
Store past inputs in array
add list command to print all past inputs

ngxingyu committed:
Fix bug where only a single word is passed into the callback method.

ngxingyu committed:
Add Task class to represent a task.
Update Duke class to store input as tasks
Add command for toggling task done state.

ngxingyu committed:
Add ToDo, Deadline, Event subclasses of Task
Change the input handling to treat all input as a single word command followed by an arbitrary input string.

ngxingyu committed:
Properly strip all input strings
add test cases

ngxingyu committed:
Implement and use DukeException

ngxingyu committed:
use ArrayList instead of array for tasks

ngxingyu committed:
Add the delete command
update tests

ngxingyu committed:
Add more tests

ngxingyu committed:
Convert commands to enum

ngxingyu committed:
Add storage class for saving and loading tasks.

- handles conversion from list of tasks to string and back.
- saves and loads tasks from data/tasks.txt.
- does nothing if file doesn't exist.

Modify Task class to support encoding/decoding Task into a string.

- update subclasses of Task.java based on updated interface.

Update Duke class to load and save tasks to disk.

- load tasks from disk on start.
- save tasks when any tasks modified/added/deleted.
- does nothing if file cannot be created.

Update runtest.bat to remove old tasks.txt file when testing.

ngxingyu committed:
Teach Duke to understand date and time command

- Update Deadline to take in date and time in specific dd/MM/yyyy HH:mm format
- update input and Expected output

ngxingyu committed:
Merge changes from Level 8:

- Use specified date format instead of String
- Modify Storage class to ignore invalid format in saved tasks.

ngxingyu committed:
Merge branch 'master' into add-gradle-support

ngxingyu committed:
Add basic gradle support
- build.gradle from cs2103t repo add-gradle-support branch.
- checkstyle files from se-education intermediate repo.

ngxingyu committed:
Improve directory structure, use packages
Update build.gradle to use updated package structure

ngxingyu committed:
Update runtest.bat to use updated packages

ngxingyu committed:
Refactor ArrayList<Task> to standalone TaskList class
- Handle storage saving / loading from files.
- handle add / delete / toggle completion for tasks.

ngxingyu committed:
Extract some display code to Ui class.

- welcome message, hi and bye message, generic loading error message.
- TODO: extract more error messages into Ui class

ngxingyu committed:
Improve the use of OOP in code

- Update Duke.java run method to follow provided format.
- Extract commands into individual classes.
- Update Ui class to add more useful display methods - e.g. showLine, showError, readCommand.
- Remove some unnecessary error handling in TaskList
- Update runtest.bat to use updated directory structure.

ngxingyu committed:
Implement unit tests for some methods.

- override equals for all commands / tasks
- fix nullptr exception in TaskList

ngxingyu committed:
Implement Find command
- implement toString for TaskList for reusability
- implement constructor for TaskList without storage

ngxingyu committed:
Reformat files based on checkstyle.

- Add Javadocs to fix checkstyle errors.
- change ui indent to use tabs.

ngxingyu committed:
Add java docs for public methods.

ngxingyu committed:
Merge JavaDoc branch into master.

ngxingyu committed:
Merge Level 9 Find.

ngxingyu committed:
Fix coding style for javadocs
- improve header comments description

ngxingyu committed:
Add basic gui using javafx.

Based on part 1 and 2 of the javafx guide.

ngxingyu committed:
Fix unicode symbol representation to be displayed on the gui.

ngxingyu committed:
Represent the user and duke speech as Message objects.

- contains the text, user (duke or user) and whether the program should be quit.
- update all commands to output this message object and remove direct ui manipulation from commands.

ngxingyu committed:
Add gui using fxml following the javafx tutorial part 4

Add Launcher as the main entry point to the program

Remove run loop from Duke, and replace with callback function that sends message in Ui.

darrtzj commented on src/main/java/duke/Message.java:
```diff
@@ -0,0 +1,30 @@
+        DUKE
+    }
+
+    /**
+     * Creates a message with the given text, exit status, and user.
+     * @param message message to be displayed
+     * @param shouldExit whether the program should exit
+     * @param user user who sent the message
+     */
```
Javadoc comments require punctuation behind each parameter description, I noticed the same issue in several other places too.

darrtzj commented on src/main/java/duke/Message.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    /**
+     * Creates a message with the given text, exit status, and user.
+     * @param message message to be displayed
```
Javadoc comments require empty line between description and parameter section, i noticed the same issue in several other places too

darrtzj commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,135 @@
+    public Storage(String fileName) throws DukeException {
+        String directoryPath = null;
+        try {
+            directoryPath = Path.of(this.getClass()
+                    .getProtectionDomain()
+                    .getCodeSource()
+                    .getLocation().toURI()).toString();
```
I like how you introduced line break here to improve readability.

darrtzj commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.Timer;
+import java.util.TimerTask;
+
+import duke.ui.MainWindow;
+import javafx.application.Application;
```
Might be good to add an empty line to separate the import from duke and javafx?

darrtzj commented on src/main/java/duke/ui/MainWindow.java:
```diff
@@ -0,0 +1,67 @@
+package duke.ui;
+
+import duke.Duke;
+import duke.Message;
+import javafx.fxml.FXML;
```
Might be better to add an empty line between duke and javafx import statements?

darrtzj reviewed (commented):
Looks good. Some minor Javadoc issues which can be fixed pretty quickly.

ngxingyu committed:
Fix issues not adhering to coding standard.

- javadoc statements should include newline after description and end with period
- add space between imports from different packages.

ngxingyu committed:
Fix coding standard issues with import order and separation.

ngxingyu committed:
Fix checkstyle issues in test files.

ngxingyu commented:
> Looks good. Some minor Javadoc issues which can be fixed pretty quickly.

Thanks for you feedback, they were very useful in clarifying my understanding of the coding standard.
The only point I did not modify is the one on adding an empty line between the duke and javafx imports as I believe the coding standard doesn't require this and checkstyle flags out the empty line if I include it.

loyhongshenggg commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Creates a duke chatbot which persists tasks to the file named fileName.
+     *
+     * @param fileName      Name of the file to store tasks in.
```
Is there a reason for the spacing next to fileName? 😅

loyhongshenggg commented on src/main/java/duke/Message.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class Message {
+    public final String message;
+    public final boolean shouldExit;
```
Good job on the boolean value 'shouldExit'!

loyhongshenggg commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,47 @@
+ * The Duke chatbot
+ */
+public class Duke {
+    private final TaskList tasks;
```
For constants, perhaps you may want to consider having them in full caps (so that it is easier to distinguish)

loyhongshenggg commented on src/main/java/duke/commands/InvalidCommand.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    public Message execute(TaskList tasks) throws DukeException {
```
The name 'execute' might be better to state 'executeSomething' for example

loyhongshenggg commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,139 @@
+     *
+     * @return the type of the task.
+     */
+    public abstract Type getType();
```
For getters/setters, keep in mind that header comments can be omitted 🤗

loyhongshenggg reviewed (commented):
Overall very consistent and good effort on following the code standards! Very neat with extra considerations, out of the coding standards, to improve readability!

Here are a few things to note:
1. Headers can be omitted for getters and setters! (save your time haha)
2. For methods, as far as possible, avoid having one word verbs like execute. Can be ambiguous.

ngxingyu committed:
Add some assertions to Duke and Parser

Code currently contains no assertions.

It needs to change as using assertions is one of the project requirements and some of the assertions ensure assumptions that parameters are non-null might be useful.

Adding these assertions let's us check that we have handled the null cases for filename and provided a non-null consumer to Duke.

ngxingyu committed:
Update build.gradle main class to point to Launcher.

ngxingyu committed:
Setup CI based on example.

ngxingyu committed:
Improve coding style.

- Rename directory variable within Storage class for clarity.
- Remove some duplicate code, reduce nesting in Parser.

ngxingyu committed:
Fix bug with invalid datetime.

- catch datetime parse exception error and return InvalidCommand

ngxingyu committed:
Merge pull request #1 from ngxingyu/heads/A-Assertions

Add some assertions to Duke and Parser

ngxingyu committed:
Merge branch 'master' into A-CodeQuality

ngxingyu committed:
Detect duplicate tasks in list on add.

- Uses the equals method to compare task.
- throws error if tasks exists.
- Adds a containsTask method to check for task in task list.
- Add basic tests for containsTask method.

ngxingyu committed:
Merge pull request #2 from ngxingyu/A-CodeQuality

A code quality

ngxingyu committed:
Improve gui.

- differentiate messages by type, user.
- update overall layout for usability.

ngxingyu committed:
Make happy path more prominent.

ngxingyu committed:
Merge branch 'A-CodeQuality'

ngxingyu committed:
Merge branch 'master' of github.com:ngxingyu/ip

ngxingyu committed:
Use Optional to handle null cases instead of throwing.

Refactor TaskList for clarity.

- Extract method to save tasks to storage
- improve happy path in removeTask method.

ngxingyu committed:
Merge branch 'branch-A-CodeQuality'

ngxingyu committed:
Fixed null-pointer exception regression introduced in ef45d496

- check that storage is non-null

ngxingyu committed:
Merge branch 'Gui-Improvements'

ngxingyu committed:
Handle more issues with code standard

Add variant of addTask method that does not throw exception.

ngxingyu committed:
Update duke image.

ngxingyu committed:
Add more tests.

1. Add tests for all commands.
- Fix some missing `equals` and `hashCode` methods
- Fix `equals` implementation in `ExitCommand`

2. Add Mockito as a dependency for mocking TaskList behavior

3. Test loading and saving of tasks to file.

ngxingyu committed:
Update docs README with features in App

Update Ui screenshot.
